src/libfsm/collect.c

```diff
@@ -0,0 +1,108 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include <adt/queue.h>
+#include <adt/set.h>
+#include <adt/edgeset.h>
+#include <adt/stateset.h>
+
+#include "internal.h"
+
+static void
+clear_states(struct fsm_state *s)
+{
+	for (; s != NULL; s = s->next) { s->reachable = 0; }
+}
+
+static int
+mark_states(struct fsm *fsm)
+{
+	const unsigned state_count = fsm_countstates(fsm);
+	struct queue *q = queue_new(state_count);
+	int res = 0;
+	if (q == NULL) { return 1; }
+
+	if (!queue_push(q, fsm->start)) {
+		goto cleanup;
+	}
+	fsm->start->reachable = 1;
```

Comment:
I'll move this to the predicate bitmap, when I'm closer to making a PR for that.